### ISSUE TITLE
Containers stopped working in Firefox 53.0

### ISSUE THREAD

**Issue by pwaring** [mozilla/testpilot-containers#469]
I've been using the Containers feature for a few weeks in Firefox 52 (using the `userContext` options in `about:config`). It worked fine until Firefox auto-updated to version 53 -- now File -> New Container Tab is greyed out and a long-click on the new tab button no longer gives me the container options.

This has happened in Windows 8, Ubuntu 16.10 and macOS 10.12.4 so it appears to be a generic issue rather than platform-specific. The functionality was definitely working in 52.0.2 on all platforms and stopped working in 53.0 so appears to be a regression bug.

The relevant `about:config` settings are:

 * `privacy.userContext.enabled`: User set to `true`
 * `privacy.userContext.longPressBehavior`: Kept as default of `0` (this option didn't exist when I originally enabled containers)
 * `privacy.userContext.ui.enabled`: User set to `true`
 * `privacy.usercontext.about_newtab_segregation.enabled`: User set to `true`

**Comment by bakulf** [comment]
Does it happen also if you install the test-pilot container addon?

**Comment by pwaring** [comment]
I've just installed test-pilot and restarted Firefox. The menu item is still disabled and long-clicking the new tab icon doesn't work.

**Comment by jonathanKingston** [comment]
@pwaring could you paste the value of: "browser.privatebrowsing.autostart" and also the extensions you have in about:support?

Thanks

**Comment by jonathanKingston** [comment]
`privacy.userContext.ui.enabled` should be set to `false` also

**Comment by jonathanKingston** [comment]
@pwaring the long press should be replaced with a hover menu on the + icon instead in the test pilot.

**Comment by pwaring** [comment]
`browser.privatebrowsing.autostart` is set to true (I haven't set this directly but I assume it's a consequence of having Firefox start in private browsing mode by default).

`privacy.userContext.ui.enabled` is now false (I didn't change it, I assume test-pilot did).

If I hover over the + icon I get a tooltip saying 'Open a new tab (Ctrl + T)'.

**Comment by jonathanKingston** [comment]
> `browser.privatebrowsing.autostart` is set to true (I haven't set this directly but I assume it's a consequence of having Firefox start in private browsing mode by default).

This is the problem, currently we don't support that. We should have put a warning in about this. However it's actually so infrequently used you are the first to report it.

The following setup gets you 90% there instead (this is what I use):
![selection_612](https://cloud.githubusercontent.com/assets/338988/25532024/a5829762-2c23-11e7-9bc1-368e59a080ed.png)

**Comment by pwaring** [comment]
What I don't understand though is why it stopped working in 53 - I had the same settings in 52.02 so containers did work with private browsing on by default. Can you not revert whatever change broke the functionality?

**Comment by jonathanKingston** [comment]
@pwaring what was your setup in 52, were you using the test pilot experiment or the preference setup?

I made the assignment feature depend on private mode being off, however no other changes were made to depend on this setting so it shouldn't have made a difference. Did you change the setting and not restart your browser perhaps?

The preference only version of containers hasn't ever worked for this mode (I have a patch to put in core firefox still waiting to land).

**Comment by pwaring** [comment]
I was using the preference in 52 - I only installed the test pilot experiment because @bakulf asked me to. In fact checking my Twitter history it looks like you were the person who told me how to use preferences. :)

I shutdown my PC every night and I've been using this functionality since 5th April so it's not a case of changing settings and forgetting to restart.

> The preference only version of containers hasn't ever worked for this mode (I have a patch to put in core firefox still waiting to land).

That doesn't tally with my experience. All my Firefox instances use private browsing by default and containers were working in all of them until 53 landed.

**Comment by jonathanKingston** [comment]
>  In fact checking my Twitter history it looks like you were the person who told me how to use preferences. :)

Heh yeah I thought I recognised you :bowtie:

> That doesn't tally with my experience. All my Firefox instances use private browsing by default and containers were working in all of them until 53 landed.

Yeah this comment was specific to this experiment only... I didn't think it was our extension causing this as there isn't much code checking it.

As for 53, I can only think we had bugs that showed some of these menu items in private mode when they shouldn't have been shown. The reason this bug exists is due to us hiding containers features when users are in normal private mode. However this means users like yourself have a hidden menu.

@bakulf and @kjozwiak do you know of any instances in 53 Firefox where we fixed a private menu that shouldn't have been shown?

@pwaring I can look into getting my core patch landed to enable containers again in alwaysPrivateMode which would likely give you these menus back in 55. Unfortunately small regressions like this especially for not core functionality of Firefox wouldn't be able to be backed out.

**Comment by pwaring** [comment]
I don't know if something has changed but I can now get a container by clicking the Containers icon on the menu in 53.03. It's not quite as convenient as the new tab button and it insists on telling me about containers every time I restart Firefox, but I can at least get at the functionality.

**Comment by jonathanKingston** [comment]
@pwaring so you think this can be closed? We have #490 too if you could try that STR that would be great.

**Comment by pwaring** [comment]
I would prefer that the functionality was restored to what it was in 52.0, since clicking on the new tab icon was more convenient and user-friendly than having to click through 3 pop-ups on the container menu.

**Comment by jonathanKingston** [comment]
@pwaring this is still the experience in Firefox itself with the containers preference `privacy.userContext.enabled` set to true in about:config

**Comment by pwaring** [comment]
I'm not sure what you mean by that. If I try using 53.0.2 with the various settings in `about:config` then I can't create a new container tab unless I install the test pilot add-on, and then only by going through three reminders each time I restart my browser.

**Comment by jonathanKingston** [comment]
Without the test pilot you can add/remove containers in about:preferences
under privacy.

**Comment by pwaring** [comment]
Yes, I can change the list of containers. What I can't do without the Test Pilot add-on is open a new tab in a given container. There is no long-click option on the New Tab icon and the New Container Tab option in the File menu is greyed out.

**Comment by groovecoder** [comment]
> There is no long-click option on the New Tab icon and the New Container Tab option in the File menu is greyed out.

Are you in a Private Browsing window by any chance?

**Comment by pwaring** [comment]
Yes, hence the title of the issue...

**Comment by yareckon** [comment]
Just adding that @pwaring is not the only one wanting this feature to work correctly in Permanent Private Browsing Mode.

**Comment by jonathanKingston** [comment]
The comments in #548 suggest that this was a change in Firefox and the time difference is release channels perhaps? I suspect it was a 53 change that did this and outside what we can control in the experiment.

**Comment by pwaring** [comment]
Update: I can no longer open new container tabs at all (with the add-on), as of 54.x (I think - I'm on 54.0.1). If I click the container link and then choose a container, nothing happens.

I can still add/edit existing containers but without the ability to use them this feature is unusable.

**Comment by pwaring** [comment]
This still seems to be broken in Quantum (57.0) - the New Container Tab is still greyed out.

**Comment by adisandro** [comment]
Voting for this feature to be implemented.

**Comment by wiktorn** [comment]
I also vote for this feature. Is there any other way to vote? Can't see anywhere documented what's the best way to vote for feature.

**Comment by pwaring** [comment]
I can't see any way to vote either - obviously as the original reporter I would like this to be fixed. :)

**Comment by voidstarstar** [comment]
@wiktorn @pwaring I think you're expected to click :+1: on the original post in order to "vote" for it.

When in a Private Browsing window, if I click the button to create a new tab in a container, I get the following error message:
`Error: Illegal to set non-private cookieStoreId in a private window`

A similar experience appears to be noted here: https://github.com/mozilla/multi-account-containers/issues/1164#issuecomment-376662522

Is this related to the new contextIdentities API?  It sounds like MAC is attempting to create a new non-private cookieStoreId from within a private window.  Is it possible to create more than one private cookieStoreId?

If there's a default non-private container, I wonder if there should also be a default private container.

**Comment by voidstarstar** [comment]
It looks like this depends on this bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1320757

This issue specifically talks about permanent private browsing mode, but I see no reason why this shouldn't be permissible for regular private browsing mode too.  Most use cases for permanent should also apply to non-permanent.

Non-permanent private browsing example:
A user has 3 websites (A, B, and C) that they visit, but does not want any of them sharing a session, so they'd like to use containers.  They open Website A in a normal non-private window because they want to save history.  They press Ctrl+p and open Websites B and C in a private browsing window because they don't want them saving history.  Currently there's no way for B and C to use private browsing while not sharing the same session.

**Comment by voidstarstar** [comment]
@jonathanKingston I just realized that you were the author of [a patch for this](https://phabricator.services.mozilla.com/D11453).  Maybe I'm misunderstanding your patch, but it looks like it is trying to allow containers to work in private windows when `permanentPrivateBrowsing` is enabled but not work in private windows when `permanentPrivateBrowsing` is disabled.  Is there a reason why containers shouldn't just be enabled universally?  You stated:
> We disabled containers for when in private mode due to the confusion caused by having both personal container open and private personal container open.

So is the reason for this limitation simply to prevent a user from seeing the same container being used in both a private and non-private window?  If so, this sounds like a UX issue rather than a technical one.  Instead of a blanket ban of containers for all private windows, I can think of 2 possible solutions:
1. Do not allow the same container to be simultaneously used in both a private and non-private window.  If a user has a private and non-private window open, they must use 2 containers that each have 1 session.  There would be a default normal container and a default private container.
2. Allow the same container to be simultaneously used in both a private and non-private window, but clearly differentiate between them with a visual indicator (e.g. a purple mask).  If a user has a private and non-private window open, they can have 1 container that uses 2 sessions.  There would only need to be a single default container.

I suspect that option 1 is more in line with user expectations, but option 2 is what you were thinking of when you disabled this feature.  In option 1, private browsing mode could be implemented as just a "type" of container.  All containers would have a flag that designates them as either "normal" or "private".  Then if the user enables `permanentPrivateBrowsing`, it would just open the default private container and just prevent "normal" containers from running while still permitting "private" ones.  I think this approach would give users the most freedom while still making the feature intuitive to use.

**Comment by pwaring** [comment]
I think there are two different use cases for private browsing and containers, albeit with a bit of overlap. For example, the biggest selling point of containers to me is not to do with privacy or isolating sites such as Facebook, but the ability to have simultaneous sessions with the same website - I think containers are the only way to achieve that within the same browser.

**Comment by LoZio** [comment]
Adding to the voters here! I just switched to Firefox from Chrome just to use the containers feature. I usually have no history recorded and discovered the feature is not available.
On my part, I see more confusion installing an extensione (fb containers) or reading docs about containers and then not having the way to use it without apparent reason, than having them both working. At least we need some type of message telling you what's happening, but I would greatly prefer to have the feature.

**Comment by voidstarstar** [comment]
It should be noted that starting in Firefox 67, extensions will be disabled in private browsing by default.  This means that any potential confusion caused by private browsing would have required a user to manually enable this extension for private browsing.  Since a user now explicitly needs to enable this, I think the risk of confusion is greatly diminished.

**Comment by Utini2000** [comment]
I also vote for this. I use firefox at several clients/PC's in permanent private mode and it's sad to see that this extension doesn't work in such an environment. 
Ofcourse I enabled the extension to work in private mode.

**Comment by opusforlife2** [comment]
First of all, with all the similar terminology floating around here, I'm not even sure this is the issue I need to be commenting on. Can someone confirm? I have "Use custom settings for history" selected, and "Always use private browsing mode" enabled under that.

If I'm at the right place, then I add my vote. I find it a bit strange that this extension wasn't introduced with private browsing support from the start. A user who is privacy conscious to the level of using private browsing permanently would definitely want to have container support.

**Comment by pwaring** [comment]
@opusforlife2 Yes, if you have those settings ("Always use private browsing mode") then you will be affected by this issue.

**Comment by pwaring** [comment]
I'm going through some of my outstanding issues and noticed that this one has been open for four years now without a resolution (which isn't a criticism, it's an open source project at the end of the day and I certainly don't have the time or skills to fix it). Is it worth closing this issue to keep the number of open issues manageable?

**Comment by opusforlife2** [comment]
I think the maintainers will do it on their own if there is such a need.

Meanwhile, it's better if this stays open, if only for the reason that duplicates aren't opened. And if by some miracle, a person with the knowledge to fix this appears, it can only help.

**Comment by grahamperrin** [comment]
> ## … permanentPrivateMode

⚠ Side note, from <https://bugzilla.mozilla.org/show_bug.cgi?id=1521372#c6> (… not permanent private browsing): 

> the second window (the window that is opened with Control-N) truly **lacks privacy**

**Comment by grahamperrin** [comment]
![image](https://user-images.githubusercontent.com/192271/115954225-b8d06d80-a4e7-11eb-8546-231ffd780c6a.png)

https://github.com/mozilla/multi-account-containers/issues/469#issuecomment-826059683

Given (1) the fundamental necessary technical divide between Firefox containers and Firefox private browsing, and (2) the recent privacy-related enhancements to 'non-private' browsing with strict ETP: 

* I vote to close. 

https://github.com/mozilla/multi-account-containers/issues/469#issuecomment-826061907

> … better if this stays open, if only for the reason that duplicates aren't opened. …

Whilst I applaud the wishful thinking, the reality in a repository such as this is that the _open_ state does not effectively reduce duplication. 

Attempting to work around the duplication that exists can be quiet nightmare. Respectfully: things that linger open _with little or no possibility of resolution_ only add to the frustration and increase the likelihood of duplication. 

----

I'll follow up with a suggestion that may make closure more palatable.

**Comment by grahamperrin** [comment]
## Workflow suggestion

### Preparation

1. If your `default` profile has `browser.privatebrowsing.autostart` `true`, make it `false`
2. about:profiles
3. create a non-default profile `permanentprivate`
4. launch the profile
5. about:config
6. `browser.privatebrowsing.autostart` `true`
7. add [Open With](https://addons.mozilla.org/addon/open-with/)
8. set the browser preference that's pictured below
9. quit
10. add [Always In Container](https://addons.mozilla.org/addon/always-in-container/) to your `default` profile
11. also, if you like, add Open With. 

![2021-04-24 11:04:59](https://user-images.githubusercontent.com/192271/115955348-d1438680-a4ed-11eb-8a93-48ec71c12823.png)

### Workflow demonstration

Ignore the first ten seconds …

https://user-images.githubusercontent.com/192271/115955489-a4dc3a00-a4ee-11eb-8155-d8dd5bec2db7.mp4

### Variations

If you prefer: 

* make the 'permanentprivate' profile your default; alter your preparation accordingly.

**Comment by voidstarstar** [comment]
I don't think that this should be closed.  It's a very popular feature request without a full workaround.

Even though some new features of private browsing were made available for non-private browsing, there are still others that were not (e.g. temporary data is only memory resident).  Private browsing clearly still has a use case or else it would have been deprecated.


There's still an open issue for this in bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1320757

It seems that a patch was created, but the issue was unassigned and the patch was abandoned.  (It doesn't appear that the author plans on working on this and tried to pass the torch to someone else.  I may be wrong, but I don't think they are currently working on containers/firefox anymore.)

I have not tested the patch, so I don't know if it is complete/working.  It looks relatively straight forward, so I'm not sure what other technical hurdles there are if any.  If it was just overlooked, then hopefully it will be reassigned.

**Comment by christian179717** [comment]
Will it be implemented soon?
